### DISCUSSION TITLE
Issues with METplus Install (possibly with build step?)

### DISCUSSION THREAD

**Discussion by jhicks14** [dtcenter/METplus#1149] [Installation] [answered]
Hello,

I am trying to install the latest versions of METplus/MET locally. I have been using a shared NOAA install, but I am interested in downloading/installing to my local NOAA machine so that I have more control over updates and such. 

If followed the install procedures on the installation guide: https://metplus.readthedocs.io/en/latest/Users_Guide/installation.html, but I am not having any luck running METplus after completing all steps. I've changed the MET_INSTALL_DIR, CONFIG_DIR, and INPUT_BASE in my configuration file (otherwise it's the same config file I used for the original METplus install), but running run_metplus.py begins to run METplus and then freezes. I believe the issue may lie with the build script, but I'm not entirely sure (there seem to be a few errors early in the build run, but they may be harmless?). 

I tried re-installing, but still had no luck.

**Comment by jprestop** [comment]
Hi @jhicks14.  I see you are having trouble with the installation of the METplus and MET packages.

In the link to the [METplus Installation Guide](https://metplus.readthedocs.io/en/latest/Users_Guide/installation.html), I see it leads to the [Software Installation section](https://met.readthedocs.io/en/latest/Users_Guide/installation.html#software-installation-getting-started) of the MET User's Guide, where it says "Users can find the script and its instructions under on the [Downloads](https://dtcenter.org/community-code/model-evaluation-tools-met/download) page of the MET website."  

Can you please confirm that it is the compile_MET_all.sh script from the Downloads page that you are using to install MET?

**Comment by jprestop** [comment]
HI @jhicks14.  A colleague of mine just pointed out some instructions further down on the METplus Installation Guide:
https://metplus.readthedocs.io/en/latest/Users_Guide/installation.html#build-components-and-using-manage-externals

This leads me to believe you were using the build_MET.sh script for installation instead of the compile_MET_all.sh script.  MET is intended to be compiled by users from a tarfile, not directly from the repository, which is why our [Downloads](https://dtcenter.org/community-code/model-evaluation-tools-met/download) links directly to the tarfile on [GitHub](https://github.com/dtcenter/MET/releases/download/v10.0.0/met-10.0.0.20210510.tar.gz) that users should compile. We do development within the repository but run a script to create a release. That script runs some tools to automatically generate some code for utilities that are needed for development but not for compiling the released versions. 

To compile MET, please refer to the section titled:
**Sample Script For Compiling External Libraries And MET**

On the MET downloads page:
https://dtcenter.org/community-code/model-evaluation-tools-met/download

We realize that the presence of the build_MET.sh script in the METplus repository is confusing. It is included as part of our work toward the use of manage_externals, but unfortunately, it isn't working as expected. In fact, that script does just call the compile_MET_all.sh script I've referenced above. Ideally, that would all just work out of the box, but it looks like we'll need to work on refining it some more.

**Reply by jhicks14** [reply]
Good morning,

Thanks for the recommendations! I think I almost have this fully figured out. The only thing that seems to be missing after using the compile_MET_all.sh script (and passing in a file with the environmental variables) is the /bin directory that contains the MET tools such as plot_data_plane. 

Does this indicate that I set my environmental variables incorrectly?

**Comment by jprestop** [comment, marked as answer]
Hi @jhicks14.  It's hard to say why the executables were not built. It would be helpful to look at your config.log file and a make.log file, which should be in your met-10.0.0 directory, and also your configuration file for the compile_MET_all.sh script.   Could you please send those to us, either by attaching them here (if it won't let you attach the configuration file for the compile_MET_all.sh script because of file extension, you can just add a .txt to the end of the filename) or by following the "How to Send Us Data" section of the [Resources for Troubleshooting](https://github.com/dtcenter/METplus/discussions/954) post to send them via ftp?  Thanks!

**EDIT:  The user sent data.  Unfortunately, this issue was not reproducible and has not been encountered previously.  It was suggested that the user contact the NetCDF helpdesk for support.  See the comments below for details, however, the user decided to produce a different route.**

**Reply by jhicks14** [reply]
Thanks again for walking me through this! I've attached the three files you asked for. 

[config.log](https://github.com/dtcenter/METplus/files/7145832/config.log)
[make.log](https://github.com/dtcenter/METplus/files/7145833/make.log)
[install_met_env.jhicks.txt](https://github.com/dtcenter/METplus/files/7145836/install_met_env.jhicks.txt)

**Reply by jprestop** [reply]
Thank you for sending these!  They were helpful.

From your make.log file I see that the necessary “netcdf” file cannot be found.

```
In file included from nc_var_info.cc:24:0:
nc_utils.h:21:18: fatal error: netcdf: No such file or directory
 #include <netcdf>
```

From your config.log file, I see that you have the following set:

```
MET_NETCDF='/export/cpc-lw7-jhicks/jhicks/MET/METv10.0/external_libs'
MET_NETCDFINC='${MET_NETCDF}/include'
MET_NETCDFLIB='${MET_NETCDF}/lib'
```

It appears that NetCDF was not installed in /export/cpc-lw7-jhicks/jhicks/MET/METv10.0/external_libs as expected.  Do you happen to have the screen output from the compile_MET_all.sh script that you could send?  If not, could you please rerun the script and gather the screen output?  It would be helpful to take a look at that to see where the problem starts.

**Reply by jhicks14** [reply]
Sure thing! I'll attach the screen output here.
[compile_output.txt](https://github.com/dtcenter/METplus/files/7172472/compile_output.txt)

**Reply by jprestop** [reply]
Thank you, @jhicks14.  That is very helpful.  In the screen output, I can see that there was a problem compiling NetCDF-CXX:

`configure returned with non-zero (1) status`

Could you please send the config.log file and make.log file from the NetCDF-CXX directory: /export/cpc-lw7-jhicks/jhicks/MET/METv10.0/external_libs/netcdf/netcdf-cxx4-4.3.0 so I can take a look and try to figure out what's going on?

Also, while the output of the compilation of BUFRLIB did not show an error, I suspect there will be an error based on another [Discussion post](https://github.com/dtcenter/METplus/discussions/1157) that I just responded to earlier today.  I see in your config.log file that you are using 4.8.5 of the GNU compilers.  If you look in the make.log file for BUFRLIB, I suspect you will find the following message:

`gfortran: error: unrecognized command line option ‘-fallow-argument-mismatch’; did you mean ‘-Wno-argument-mismatch’?`

‘-Wno-argument-mismatch' was deprecated in favor of ‘-fallow-argument-mismatch'. It looks like that was deprecated with version 10 of the GNU compilers (https://gcc.gnu.org/gcc-10/changes.html).

Let's have you try changing the following line in the compile_MET_all.sh script from:

${FC} -c -fno-second-underscore **-fallow-argument-mismatch** \`./getdefflags_F.sh\` modv*.F moda*.F \`ls -1 *.F *.f | grep -v "mod[av]_"\` >> make.log 2>&1

to 

${FC} -c -fno-second-underscore **-Wno-argument-mismatch** \`./getdefflags_F.sh\` modv*.F moda*.F \`ls -1 *.F *.f | grep -v "mod[av]_"\` >> make.log 2>&1
 
Then, please try recompiling everything and seeing if that does the trick for BUFRLIB.  If you'd like to wait to recompile BUFRLIB until I'm able to give more feedback on the NetCDF-CXX issue, that would be fine.  I have a [GitHub issue](https://github.com/dtcenter/MET/issues/1395) set up to make improvement to the script, and I have just added this as a change to improve the script.  (If GNU version 10 or greater, have ‘-fallow-argument-mismatch' in BUFRLIB compilation. If less than 10, have ‘-Wno-argument-mismatch'' in BUFRLIB compilation.)

**Reply by jhicks14** [reply]
I'm attaching the config.log file from the netcdf directory located within the external libraries. I don't see a make.log file, however. 
[config.log](https://github.com/dtcenter/METplus/files/7181564/config.log)

Upon looking at the BUFRLIB make.log file, I do get the error "gfortran: error: unrecognized command line option ‘-fallow-argument-mismatch’". I made changes to the compile.MET.all.sh script and recompiled, and this got rid of the warning in the BUFRLIB make.log file.

**Reply by jprestop** [reply]
Thank you for sending the NetCDF-CXX log file.  I see the following:

```
/usr/bin/ld: cannot find -lnetcdf
collect2: error: ld returned 1 exit status
```
Perhaps NetCDF-C had problems compiling.  If it compiled successfully, you should see the following files in the following directories:

/export/cpc-lw7-jhicks/jhicks/MET/METv10.0/external_libs/include/netcdf.h
/export/cpc-lw7-jhicks/jhicks/MET/METv10.0/external_libs/lib/libnetcdf.a
/export/cpc-lw7-jhicks/jhicks/MET/METv10.0/external_libs/lib/libnetcdf.so

Could you please send me the config.log file and the make.log file for NetCDF-C, which I believe will be in /export/cpc-lw7-jhicks/jhicks/MET/METv10.0/external_libs/netcdf/netcdf-4.4.1.1/?

Thank you!

**Reply by jhicks14** [reply]
I see this file: /export/cpc-lw7-jhicks/jhicks/MET/METv10.0/external_libs/include/netcdf.h
But the two files within the /lib directory (libnetcdf.a and libnetcdf.so) are not there.

Here are the config.log and the make_install.log files for NetCDF-C: 
[make_install.log](https://github.com/dtcenter/METplus/files/7187096/make_install.log)
[config.log](https://github.com/dtcenter/METplus/files/7187097/config.log)

Thanks again for the assistance!

**Reply by jprestop** [reply]
Unfortunately, this one has me a bit stumped.  I haven't seen this problem with NetCDF-C previously.  In your make_install.log file I see:

```
In file included from ncx.h:11:0,
                 from v1hpg.c:13:
v1hpg.c: In function ‘ncx_len_NC’:
v1hpg.c:1276:25: error: ‘NC_64BIT_DATA’ undeclared (first use in this function)
```

NC_64BIT_DATA is defined in netcdf.h:

```
netcdf.h:138:#define NC_64BIT_DATA    0x0020  /**< CDF-5 format: classic model but 64 bit dimensions and sizes */
netcdf.h:139:#define NC_CDF5          NC_64BIT_DATA  /**< Alias NC_CDF5 to NC_64BIT_DATA */
```

But somehow it is not being located.  

The only difference in the "libtool: compile:" line just above the error in your make_install.log file and my make_install.log with a successful compilation is that you have two paths being included in addition to the path with external_libs:

```
-I/nawcf/home/nawcf1/v7.14.0/gempak/include 
-I/nawcf/home/nawcf1/v7.14.0/os/linux3.10.0_x86_64/include
```

Please try removing those paths from CFLAGS and FFLAGS and then try rerunning the script.  I see you currently have the following set:

`CFLAGS='-I/nawcf/home/nawcf1/v7.14.0/gempak/include -I/nawcf/home/nawcf1/v7.14.0/os/linux3.10.0_x86_64/include -DLinux -DUNDERSCORE -I/usr/X11R6/include   -DG_64BIT'`

`FFLAGS='-I/nawcf/home/nawcf1/v7.14.0/gempak/include -I/nawcf/home/nawcf1/v7.14.0/os/linux3.10.0_x86_64/include -fno-range-check -fno-second-underscore -ffixed-line-length-none -ff2c -std=legacy'`

I am not familiar with the -DG_64BIT option either, but you might also try removing that option from CFLAGS.

Please let us know how it goes.

**Reply by jhicks14** [reply]
I'm a bit confused which file I'm supposed to remove the CFLAGS/FFLAGS paths from, as I don't see these in any file (I'm probably missing it though)

**Reply by jprestop** [reply]
Hi @jhicks14.  I can't be sure where those are being set, but my guess would be either in a .cshrc or .bashrc file.  In any case, you could add the following to your install_met_env.jhicks file to override the current values that are set:

```
export CFLAGS='-DLinux -DUNDERSCORE -I/usr/X11R6/include'
export FFLAGS='-fno-range-check -fno-second-underscore -ffixed-line-length-none -ff2c -std=legacy'
```

Then, the compile_MET_all.sh script should use those new values from your configuration file.  I hope that helps!

**Reply by jhicks14** [reply]
Ahhh I see! I added the CFLAGS and FFLAGS lines to my install_met_env.jhicks file, but unfortunately I still get the "‘NC_64BIT_DATA’ undeclared" error.

There may be a sample install_env file that was used for the shared install of MET. I'll try to use a similar version of that file to see if the configuration is successful.

**Reply by jprestop** [reply]
@jhicks14 Thanks for trying that out. I thought there was a chance that another older netcdf.h file was in one those paths, but that seems not to be the case.  As I mentioned, I'm a bit stumped at this error.   It seems NC_64BIT_DATA can't be found even though it is in an include file that is included in this package.  

As I mentioned before, I am a bit stumped by this problem. I have not encountered it previously and can't seem to find any help inmy online searches.  Since I am not sure how to proceed, I would like to ask you to write to the helpdesk for NetCDF (support-netcdf@unidata.ucar.edu).  It seems they provide good support for their packages, and I believe they should be able to help you resolve this problem.  Could you please follow up here with their guidance and let us know how they suggest resolving this problem?  If you have any questions on how to proceed with any guidance they give since the building of NetCDF is in the script, please don't hesitate to reach out here with any questions.  Hopefully, once we can get NetCDF-C to compile, NetCDF-CXX will compile well and then MET will compile well.

**Reply by jprestop** [reply]
Hi @jhicks14.  I just wanted to follow up and see if you were able to write to the NetCDF helpdesk and if you had any luck?

**Reply by jhicks14** [reply]
Still waiting for a reply back. I'm going to try to reach out again and see if they get back soon

**Reply by jprestop** [reply]
@jhicks14 I'm sorry to hear that.  I hope you'll hear back soon.  Thank you for the update.

**Reply by jprestop** [reply]
Hi @jhicks14.  I just wanted to follow up and see if you received any response from the NetCDF helpdesk and have had any luck?

**Reply by jhicks14** [reply]
I heard back from the helpdesk, and they needed a little bit more information from me. I sent them the files they needed, but I haven't heard back.

I think I'm going to try installing on the the CPC Work servers instead and see if that gives me any issues.

**Reply by jprestop** [reply]
Thanks, @jhicks14! I'd like mark this discussion as being answered and "lock it" to prevent future posts. That's how we encourage users to create new discussions for new questions.  However, I'd also like to have the information from the NetCDF helpdesk posted here as well.  I think it would be good to leave this conversation open for now so that you can update it once you hear back from the NetCDF Helpdesk.  However, if you start the install on the CPC work servers and run into any problems that you'd like help with, please start a new Discussion (https://github.com/dtcenter/METplus/discussions/new).

**Reply by jprestop** [reply]
Hi @jhicks14.  I'm following up on my Unanswered Discussions tickets and thought I'd check in with you again to see if you had heard back from the Unidata folks about the NetCDF problems?

**Reply by jprestop** [reply]
Hi @jhicks14.  I just wanted to follow up on this Discussion status.  Have you had any luck with this problem?

**Reply by jhicks14** [reply]
Still working on it! Had to put this on the back-burner for now while I finish up some other projects.

**Reply by jprestop** [reply]
@jhicks14 That's no problem.  Thank you very much for the update.

**Reply by jprestop** [reply]
Hi @jhicks14. I'm guessing this work may still be on the back-burner but just wanted to check in about it, especially since I can't recreate this problem, haven't seen it before, and am stumped by it. :)

**Reply by jhicks14** [reply]
Hi @jprestop. Since I was having issues installing MET locally, we just moved a previous install to a directory that is accessible by more users. 

For version control, I think we'll get a system administrator to install the newest MET versions when needed. I think we can close this thread for now, and then I'll reach out if we have any more issues with future installs!

Thanks again!

**Reply by jprestop** [reply]
Thank you, @jhicks14.  I have modified the comment above in order to be able to choose an "Answer" for this question.  Please let me know if you'd like me to make any changes.  Otherwise, I'll go ahead and lock this discussion.
